src/SystemManager.hh

```diff
@@ -41,6 +44,21 @@ namespace gz
     // Inline bracket to help doxygen filtering.
     inline namespace GZ_SIM_VERSION_NAMESPACE {
 
+    /// \brief Helper container to keep track of
+    /// system interfaces and their parents
+    template<typename IFace>
+    struct SystemIfaceWithParent {
```

Comment:
Have you considered using a std::unordered_map to associate the parent entity and the system instead of a new data structure to hold the two pieces of information. A new data structure is usually the right thing to do, but this has added a lot of complexity with the changes in this file and the implementation of SystemContainer which can all be avoided by using a map.

Reply:
I'm trying to recall why I didn't do that. My guess was to minimize changing other data structures and the fact that `Iface*` itself is not hashable. I should probably try an `unordered_map`, but won't be able to work on this till next week.
With regards to `SystemContainer`, I initially implemented it because I needed to prevent interface pointers from changing. I believe that, if I were to insert a new item to an `unordered_map`, it might invalidate all references.

Reply:
Went for a middle ground. Sorry I had to force push.